Allow 'setup-workflows' to create a PR instead of pushing to main branch

f1ames opened the pull request:
As mentioned in https://github.com/ckeditor/ckeditor4-workflows-common/issues/6#issuecomment-745287750:

> > We may try to fiddle with bot/token permissions to see if it's possible to workaround.
> 
> At the moment, bot has `Write` access in `ckeditor4` repostiory. To be able to push despite branch protection rules, it would need to have `Admin` access which I think is rather a bad idea... It should have as little permissions as possible. So the other solution is to go with:
> 
> > If not, other solution may be to adjust the workflow (via config flag, see [#4](https://github.com/ckeditor/ckeditor4-workflows-common/issues/4)) so it creates a PR instead of pushing directly to main branch.
> 
> Let's try the later approach and see how it works.

So I went with configuration option which allows to alter the workflow to create PR instead of pushing directly to main branch. Since it doesn't make sense to keep configuration in GH secrets I have introduced optional JSON config file (#4).

The above needs proper explanation so I took this occasion to improve README file and describe all common workflows we have now.

The default flow and the one wit PR were tested in separate test repos:

* https://github.com/ckeditor/workflows-test-1
* https://github.com/ckeditor/workflows-test-2

Closes #6. Closes #4.

f1ames committed:
Make setup-workflows configurable to enable PRs.

f1ames committed:
Add workflows description to README file.

f1ames commented:
After this PR is merged we need manually update `setup-workflows.yml` in `ckeditor4` repository.

jacekbogdanski commented on workflows/setup-workflows.yml:
```diff
@@ -38,6 +38,19 @@ jobs:
+      - name: Prepare PR branch
+        if: env.HAS_CHANGES == 1
+        run: |
+          echo "BRANCH_NAME=${{ github.ref }}" >> $GITHUB_ENV
+          if [[ -f "./.github/workflows-config.json" ]]; then
+            pushAsPullRequest=$(jq ".setupWorkflows.pushAsPullRequest" "./.github/workflows-config.json")
+            if [[ "$pushAsPullRequest" == true ]]; then
+              echo "AS_PR=1" >> $GITHUB_ENV
+              echo "BRANCH_NAME=t/setup-workflows-update" >> $GITHUB_ENV
+              git checkout -b t/setup-workflows-update
+            fi
+          fi
```
Maybe we could load JSON file into the env variable as a separate, initial step, so the config is available everywhere? I was thinking about some separate step like:

```
- name: Load settings
  run: |
	echo "SETTINGS={}" >> $GITHUB_ENV
	if [[ -f "./.github/workflows-config.json" ]]; then
		echo "SETTINGS=${{ jq './.github/workflows-config.json }}" >> $GITHUB_ENV
	fi
```

and then use it everywhere as e.g. `jq -n .setupWorkflows.pushAsPullRequest` (should return "null" if nothing found). WDYT?

Also, maybe there is some better way to keep settings somehow easily accessible? Overall it should work for our case now, but any other setting will require duplicated logic for loading config which may be a bit of pain.

If we go with some good approach here, we will be able to just copy it in any action using config file.

f1ames replied:
Makes perfect sense. The best approach would be probably extracting config reading logic to separate action (so it can be called with two lines) or adding as external script to be reused.

However, as this is quite simple code I'm for adding it as initial step for now to each workflow which needs it (using copy/paste) and then we can see how problematic it is. I will also add short README section on this.

jacekbogdanski commented on README.md:
```diff
@@ -4,12 +4,64 @@ Shared CKEditor 4 GitHub workflows.
+
+### setup-workflows
+
+This is the main workflow responsible for propagating all common workflows. When it is added to any other repository it checkouts this repository and copies all common workflows from `workflows/` directory to `.github/workflows/` one. It is done once a day so any changes are propagated on a daily basis. See `workflows/setup-workflows.yml` file.
+
+It is a cron job task so will be triggered only on main repository branch.
```
Maybe we could add information when cron runs the action? (`02:00 UTC`).

We could also add this information to other actions.

jacekbogdanski commented on README.md:
```diff
@@ -4,12 +4,64 @@ Shared CKEditor 4 GitHub workflows.
+
+#### Optional secrets
+
+* `BRANCH_MAIN` - Target branch name. By default, workflow runs on main repository branch.
```
I'm wondering if it shouldn't be a part of new settings json file? We could default it into `github.ref` if not set (depending on the action). The good part is that we will be able to configure the branch per action if needed, and setting env variables is less friendly than config.

jacekbogdanski reviewed (dismissed):
Looks good, some minor changes required.

f1ames committed:
Add config loading step.

f1ames committed:
Add information about cron workflows times.

f1ames committed:
Read only related config part for given workflow.

f1ames committed:
Move updateDeps BRANCH_MAIN as config option.

f1ames commented:
I have updated [https://github.com/ckeditor/workflows-test-1](https://github.com/ckeditor/workflows-test-1) and [https://github.com/ckeditor/workflows-test-2](https://github.com/ckeditor/workflows-test-2) withe latest changes.

f1ames commented:
I see this line

```shell
pushAsPullRequest=$(echo ${{ env.SETTINGS }} | jq ".pushAsPullRequest")
```

 doesn't work as it should (see [here](https://github.com/ckeditor/workflows-test-2/runs/1666444731?check_suite_focus=true#step:9:3)) resulting in

```shell
pushAsPullRequest=$(echo {} | jq ".pushAsPullRequest")
```

so this still needs to be fixed.

f1ames committed:
Add dispatch trigger. Improve config loading.

f1ames committed:
Fix 'Prepare PR branch' step.

f1ames committed:
Move BRANCH_MAIN to config. Add dispatch trigger.

f1ames committed:
Move config read after checkout. Add '-c' for config formatting. Remove redundant 'Setup node' action.

f1ames commented:
I had some hard time providing proper loading for config files but it seems to be working now. Also, it seems adding manual dispatch for some jobs could be useful (and also helpful for testing). And so:

## update-deps

1.  Tested with no config - https://github.com/ckeditor/workflows-test-1/runs/1839790535?check_suite_focus=true#step:3:11
2.  Tested with empty config `{}` - https://github.com/ckeditor/workflows-test-1/runs/1839793100?check_suite_focus=true#step:3:11
3.  Tested with sample config `{"branchMain": "major"}` - https://github.com/ckeditor/workflows-test-1/runs/1839558113?check_suite_focus=true#step:3:11

All steps passes until NPM update since there is no `package.json` in this repo. But you can see based on branch names and steps that config was read and processed correctly.

## setup-worfklows

1.  Tested with sample config `{"pushAsPullRequest": true}` - [https://github.com/ckeditor/workflows-test-1/runs/1839877207?check_suite_focus=true](https://github.com/ckeditor/workflows-test-1/runs/1839877207?check_suite_focus=true) and PR was created (ckeditor/workflows-test-1#1)
2.  Tested with sample config but read from config file - [https://github.com/ckeditor/workflows-test-2/runs/1840137819?check_suite_focus=true#step:4:11](https://github.com/ckeditor/workflows-test-2/runs/1840137819?check_suite_focus=true#step:4:11) and PR created (ckeditor/workflows-test-2#2)
3.  Tested with direct branch push with review protection branch rules (same as `ckeditor4` repo) - [https://github.com/ckeditor/workflows-test-2/runs/1840157154?check_suite_focus=true#step:11:12](https://github.com/ckeditor/workflows-test-2/runs/1840157154?check_suite_focus=true#step:11:12)

---

So it seems everything works fine but testing it is kind of cumbersome.

f1ames commented:
As for testing workflows locally there is already mentioned [act](https://github.com/nektos/act) and some other approaches like creating and testing CI-agnostic parts (e.g. https://stackoverflow.com/a/59988803/646871)

f1ames committed:
Update README on config loading.

f1ames commented:
Ready for another review.

sculpt0r commented on README.md:
```diff
@@ -4,12 +4,110 @@ Shared CKEditor 4 GitHub workflows.
+
+#### Optional configuration
+
+* `updateDeps.branchMain` : `String` Target branch name. By default, workflow runs on main repository branch.
```
Could we replace `branchMain` settings name with something like `targetBranch`?

sculpt0r reviewed (changes requested):
Based on both provided test repos, it looks good. I'm thinking if there is a way to make an initial workflow setup, that will correspond to this repo? I mean besides manually creating an initial workflow?

I was working on some auto-test approach: https://github.com/ckeditor/ckeditor4-workflows-common/blob/testing/test.sh

Maybe it's possible to use some of these lines to create a script that will create an initial workflow on a given repository?

f1ames committed:
Rename branchMain to targetBranch config in update-deps workflow.

f1ames committed:
Rename branch var in setup-workflows.

f1ames committed:
Change the way branch name is fetched.

Since 'github.ref' returns full branch path (for example 'refs/heads/main'), PR name gets long
and confusing unnecessary. I changed it to 'git rev-parse --abbrev-ref HEAD' which retruns
simple name ('main' in this case).

f1ames commented:
I updated naming in `setup-workflows` too and changed the way branch name is fetched (see a49edcb).

sculpt0r commented:
Looks good to me.

However, I found that `setup-workflows.yml` fails when try to make PR while another PR exists. It uses the same branch name as the source branch in PR.   
 

Maybe we could force \`push\` for such a branch? It will be eventually updated to the newest changes, and we will stay with a single PR?

sculpt0r committed:
Add node instalation step.

It appears that without it, we can receive: 'npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'

sculpt0r commented:
I added node installation step. 

Without 713c49f, step 'Update NPM dependencies' ends with permission denied:  
```bash
Run npm i
npm WARN saveError ENOENT: no such file or directory, open '/home/runner/work/workflow-tests-PR-6/workflow-tests-PR-6/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/home/runner/work/workflow-tests-PR-6/workflow-tests-PR-6/package.json'
npm WARN workflow-tests-PR-6 No description
npm WARN workflow-tests-PR-6 No repository field.
npm WARN workflow-tests-PR-6 No README data
npm WARN workflow-tests-PR-6 No license field.

up to date in 0.262s
found 0 vulnerabilities

npm WARN deprecated core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.
npm WARN deprecated cross-spawn-async@2.2.5: cross-spawn no longer requires a build toolchain, use it instead
npm WARN checkPermissions Missing write access to /usr/local/lib/node_modules
npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /usr/local/lib/node_modules
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/usr/local/lib/node_modules'
npm ERR!  [Error: EACCES: permission denied, access '/usr/local/lib/node_modules'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/usr/local/lib/node_modules'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2021-02-11T12_40_23_479Z-debug.log
Error: The operation was canceled.
```

Please pull new changes.

f1ames committed:
Cancel workflow if AS_PR=true and target branch already exists.

f1ames commented:
> However, I found that `setup-workflows.yml` fails when try to make PR while another PR exists. It uses the same branch name as the source branch in PR.   
> 
> Maybe we could force `push` for such a branch? It will be eventually updated to the newest changes, and we will stay with a single PR?

I added checking if the PR branch already exists - in such cases, workflow will be simply canceled.

> I'm thinking if there is a way to make an initial workflow setup, that will correspond to this repo? I mean besides manually creating an initial workflow?

We were thinking about it on a very beginning but since it is "setup and forget' and requires minimal amount of work (copy-paste `setup-workflows.yml`), there is no sense in automating this.

f1ames commented:
@sculpt0r ready for another review.

f1ames commented:
@sculpt0r ready for another review.

sculpt0r commented:
> I added checking if the PR branch already exists - in such cases, workflow will be simply canceled.

Also sounds like a good idea :)

Now `setup-workflows.yml` stuck at `Process config` in my [test repo](https://github.com/ckeditor/workflow-tests-PR-6/runs/1922927778?check_suite_focus=true)

However, I can't figure out where is the misspell or syntax error... :thinking: Could you take a look there?

f1ames committed:
Replace elif with else.

f1ames commented:
> Now `setup-workflows.yml` stuck at `Process config` in my [test repo](https://github.com/ckeditor/workflow-tests-PR-6/runs/1922927778?check_suite_focus=true)
> 
> However, I can't figure out where is the misspell or syntax error... Could you take a look there?

I used `elif` without any condition :facepalm: Replaced with `else`, see a2e1abc.

f1ames commented:
@sculpt0r ready for another review.

sculpt0r commented:
[https://github.com/ckeditor/ckeditor4-workflows-common/pull/12#issuecomment-781135492](https://github.com/ckeditor/ckeditor4-workflows-common/pull/12#issuecomment-781135492) Seems to work right now :)  
Also, the entire action looks `canceled` which is good: https://github.com/ckeditor/workflow-tests-PR-6/actions  
However, it looks like steps after `Cancel build if update branch already exist`was executed anyway and crash on `Push changes` again: [https://github.com/ckeditor/workflow-tests-PR-6/runs/1960342487?check_suite_focus=true](https://github.com/ckeditor/workflow-tests-PR-6/runs/1960342487?check_suite_focus=true)  
Another run looks good: [https://github.com/ckeditor/workflow-tests-PR-6/runs/1960334325?check_suite_focus=true](https://github.com/ckeditor/workflow-tests-PR-6/runs/1960334325?check_suite_focus=true)

You use cancel build action. The [description](https://github.com/marketplace/actions/cancel-this-build#cancel-action) says `The cancellation might take a few seconds to process`. So it might be that few other actions will be invoked before it actually works? :thinking: So it looks like there might be a situation that some action will be invoked even if you cancel the entire run.  
  
Could you take another look into the 'canceling' method and try to improve it?

f1ames commented:
> You use cancel build action. The [description](https://github.com/marketplace/actions/cancel-this-build#cancel-action) says `The cancellation might take a few seconds to process`. So it might be that few other actions will be invoked before it actually works? So it looks like there might be a situation that some action will be invoked even if you cancel the entire run.
> 
> Could you take another look into the 'canceling' method and try to improve it?

Ok, it seems like that's the case. This cancel action is kind of a workaround, since it's not that straightforward to cancel action from within itself:

*   Step can be canceled by exiting the process but next steps will be processed normally (no matter the exit code).
*   All further steps can be conditional, like `if: env.SHOULD_CANCEL == 1` but then it bloats the code.

As for the solution to the cancel action delay, we can add wait action after - if the job needs to be cancel, the cancel step is executed and then wait step. Otherwise, those to steps will be simply skipped.

f1ames committed:
Add step which waits for cancellation.

f1ames commented:
> As for the solution to the cancel action delay, we can add wait action after - if the job needs to be cancel, the cancel step is executed and then wait step. Otherwise, those to steps will be simply skipped.

As mentioned above, I've added wait step - c93ee5b.

f1ames commented:
@sculpt0r ready for another review :crossed\_fingers:

sculpt0r commented:
> @sculpt0r ready for another review

LGTM :tada:   
  
A few runs and it ends with proper cancelation. Of course, it elongates waiting time before any results - but since it's automatic - it's fine.
